### PR TITLE
Restore epoch count

### PR DESCRIPTION
**Pull request by AntoBrandi**
When using the `--continue_path` flag, the epoch and the checkpoint are restored correctly but the counter always starts from 0.
Expected Behavior:
The counter should start from the index of the restored epoch.

**Old Behavior:**
image

**New behavior:**
image

### CODE REVIEW

**Comment by eginhard** [comment]
Thank you for the PR! That does make more sense indeed.

However, it also restores the counter when using `--restore_path` and I think in that case resetting it to zero might make more sense. I'll think about how best to handle that.

If you're interested in something similar, there is also some issue with the learning rate not being restored correctly with `--continue_path`: https://github.com/idiap/coqui-ai-TTS/discussions/326

**Comment by AntoBrandi** [comment]
Thanks for the review and the useful insights!
Just for clarification, what is the difference between `--restore_path` and `--continue-path`? When should one be used over the other?

**Comment by eginhard** [comment]
In theory they should be used as follows:

`--continue_path <PATH>`:
- To continue a training run that got interrupted, e.g. because a job failed or reached a time limit
- `<PATH>` points to a folder and Coqui will resume training from the latest checkpoint in there
- Coqui _should_ restore learning rate, optimizer state etc. to the previous values, but this might not be working correctly

`--restore_path <PATH>`:
- `<PATH>` points to a specific model checkpoint file
- To start a new training run based off the given checkpoint, e.g. to fine-tune an already trained model on a new dataset
- Learning rate etc. should not be restored from the checkpoint, generally the user would specify a new value in the config

**Review by eginhard — APPROVED** [review]
Ok, I ended up taking a deeper look into everything related to restoring models and fixed a whole range of things:
- Ensure step/epoch counts are restored correctly on `continue`, but reset to zero on `restore`.
- Adopt the standard Pytorch scheduler interface for Coqui's custom LR schedulers.
- Also store and restore the scheduler `state_dict`, but only restore optimizer and scheduler on `continue`.
- Correctly display total number of steps/epochs because they are zero-indexed.
- Log learning rates of the current step, not the next one.
